client/proposer-metrics/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Prometheus basic proposer metrics.
+
+use prometheus_endpoint::{register, PrometheusError, Registry, Histogram, HistogramOpts, Gauge, U64};
+
+/// Optional shareable link to basic authorship metrics.
+#[derive(Clone, Default)]
+pub struct MetricsLink(Option<Metrics>);
+
+impl MetricsLink {
+	pub fn new(prefix: &str, registry: Option<&Registry>) -> Self {
+		Self(
+			registry.and_then(|registry|
+				Metrics::register(prefix, registry)
+					.map_err(|err| { log::warn!("Failed to register prometheus metrics: {}", err); })
+					.ok()
+			)
+		)
+	}
+
+	pub fn report(&self, do_this: impl FnOnce(&Metrics)) {
+		if let Some(metrics) = self.0.as_ref() {
+			do_this(metrics);
+		}
+	}
+}
+
+/// Authorship metrics.
+#[derive(Clone)]
+pub struct Metrics {
+    pub block_constructed: Histogram,
+    pub number_of_transactions: Gauge<U64>,
+}
+
+impl Metrics {
+	pub fn register(prefix: &str, registry: &Registry) -> Result<Self, PrometheusError> {
+		Ok(Self {
+			block_constructed: register(
+				Histogram::with_opts(HistogramOpts::new(
+					&format!("{}_proposer_block_constructed", prefix),
```

Comment:
Why have this additional `substrate` prefix? Isn't the one initialized here https://github.com/paritytech/substrate/blob/d2bba611438f5c5c379968eb42fcc1cf5d94ecbf/client/service/src/builder.rs#L1026 enough? (Hope I am not missing something.)

Reply:
Ah, is this because it is initialized both within Polkadot as well as Substrate and thus the metric names have to differ?

Do we ever have to compare the two or aggregate on the two?

Reply:
Indeed, I overlooked this, thanks!

Reply:
> Ah, is this because it is initialized both within Polkadot as well as Substrate and thus the metric names have to differ?
> 
> Do we ever have to compare the two or aggregate on the two?

No, I think it should be fine without this